### PR TITLE
GNOME 45 support

### PR DESCRIPTION
**Pull request by jezek**
These are minimum changes needed to run on GNOME 45.
Specific changes are described in commit(s).

NOTE: I'm creating this PR as a draft, cause the latch & lock indications are broken when running this PR in GNOME 45 and I still haven't figured out why. I'm a beginner in developing & debugging extensions. If anyone knows how to fix, help is welcome.

### CODE REVIEW

**Comment by sneetsher** [comment]
Thank you @jezek for the shared effort. I will look into it. Hope they didn't change much.

File: extension.js [hunk 1]
```diff
@@ -112,69 +110,66 @@ function _a11y_mods_update(o, latch_new, lock_new) {
 // Gnome-shell extension interface
 // init, enable, disable
 
-function init() {
```
Comment (sneetsher) [inline extension.js:115]: To be moved to class constructor

File: extension.js [hunk 2]
```diff
@@ -112,69 +110,66 @@ function _a11y_mods_update(o, latch_new, lock_new) {
...
+            seat = Clutter.get_default_backend().get_default_seat();
+        } catch (e) {
+            seat = Clutter.DeviceManager.get_default();
+        };
+
+        if (seat) {
+            mods_update_id = seat.connect("kbd-a11y-mods-state-changed", _a11y_mods_update);
+        };
+
+        Main.panel._rightBox.insert_child_at_index(button, 0);
```
Comment (sneetsher) [inline extension.js:150]: Main.panel._rightBox: To be checked after migration, or go with _indicator

File: extension.js [hunk 3]
```diff
@@ -17,16 +17,14 @@
...
-const Mainloop = imports.mainloop;
+import * as Main from 'resource:///org/gnome/shell/ui/main.js';
+import GLib from 'gi://GLib';
 
-const ExtensionUtils = imports.misc.extensionUtils;
-const Me = ExtensionUtils.getCurrentExtension();
+import { Extension } from 'resource:///org/gnome/shell/extensions/extension.js';
 
 //
 const dbg = false;
```
Comment (sneetsher) [inline extension.js:30]: There few global vars here that may need to be moved inside class

File: extension.js [hunk 4]
```diff
@@ -112,69 +110,66 @@ function _a11y_mods_update(o, latch_new, lock_new) {
...
+        } catch (e) {
+            seat = Clutter.DeviceManager.get_default();
+        };
+
+        if (seat) {
+            mods_update_id = seat.connect("kbd-a11y-mods-state-changed", _a11y_mods_update);
+        };
+
+        Main.panel._rightBox.insert_child_at_index(button, 0);
+        timeout_id = GLib.timeout_add(GLib.PRIORITY_DEFAULT, 200, _update);
```
Comment (sneetsher) [inline extension.js:151]: Ok, Glib for event loop

**Review by sneetsher — COMMENTED** [review]
Thank you, I will continue working on it.

**Comment by jezek** [comment]
Just a note. What I found out (from enabled logs), that the `kbd-a11y-mods-state-changed` is not triggering, so the `_a11y_mods_update` is not called ever.

**Comment by jezek** [comment]
Pardon me my question. You wrote some review comments. I know you also wrote `Thank you, I will continue working on it.` above, but I just want to be sure that I didn't misunderstood. Are the review comments just for you, or should I implement them?

**Comment by sneetsher** [comment]
@jezek  Yeah, they are comments for myself.  Just to know where I left off and look for what next check.

By the way, It seems there are more into Gnome 45. I still couldn't get it to enable or show debug logs. I will  test upstream Gnome 45 updated demo, probable do a rewrite step by step to find what broken.

**Comment by jezek** [comment]
When testing & debugging I'm using the `env GNOME_SHELL_SLOWDOWN_FACTOR=2 MUTTER_DEBUG_DUMMY_MODE_SPECS=1024x768 dbus-run-session -- gnome-shell --nested --wayland
` command, so I don't have to restart wayland session after making changes. I found out, that after this, the extension logs are being written to the output of the command, so to get just the logs I pipe the whole output to grep and filter just the  extension logs (`env GNOME_SHELL_SLOWDOWN_FACTOR=2 MUTTER_DEBUG_DUMMY_MODE_SPECS=1024x768 dbus-run-session -- gnome-shell --nested --wayland 2>&1 | grep KMS`).

**Comment by jezek** [comment]
Hi @sneetsher , it's me again.

I've been using this extension, with this PR on Manjaro Linux, even when the latch & lock indications didn't work, cause it's more convenient, to see a modifier is pressed even when latch & lock indications aren't shown.

But, after last update, the latch& lock indications started working again. From what I deduce was a bug on gnome side, not a port bug, nor an API change.

**Comment by sneetsher** [comment]
@jezek  I could get some time to test it.

- Didn't work in Ubuntu 23.10 with gnome-shell 45.2. May be bug not yet fixed there, also it is was  an upgrade from 22.04/23.04/23.10 . I will give it another shot with clean install.
- Work perfect in Fedora 39 with gnome-shell 45.0

Thank you very much for effort. I am merging it.

**Comment by sneetsher** [comment]
@jezek  I tested Ubuntu 23.10  Vanilla installation. It works perfectly too.

**Pull request merged**